### PR TITLE
[CURL][tests] Fixed encoding of rar files to be consistent with zip etc...

### PR DESCRIPTION
**Pull request by damienruscoe**
[CURL][tests] Added tests checking for differences between URIUtils::* and CURL::* methods.

Fixed issue found in testing when a service is unavailable. Check the service is non-null before calling its methods.

## Description
The aim of this PR is to start a discussion on the differences found between the duplicate implementations available surrounding CURL objects and URIUtils utility functions. 

## Motivation and context
From my understanding; URIUtils were created to quickly test strings which are representing paths avoiding the costly process of constructing a CURL object, which involves an expensive parse function. I want to investigate the usage of CURL objects through the codebase involving the constant conversion between CURL objects -> std::string (`CURL::Get()`) when a `std::string` is required for an interface and also the parsing of `std::string` into `CURL` objects when a CURL object is required.

While these test cases are trivial I would like different implementations to return the same results.

## How has this been tested?
The tests as submitted in this PR are currently wrong. I would like to find a consensus on the correct behaviour.

## What is the effect on users?
None

## Types of change
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [X] **None of the above** (please explain below)

## Checklist:
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [X] I have added tests to cover my change
- [X] All new and existing tests passed

### CODE REVIEW

File: xbmc/utils/test/TestURIUtils.cpp [hunk 1]
```diff
@@ -600,3 +600,118 @@ TEST_F(TestURIUtils, UpdateUrlEncoding)
...
+
+  {
+    // Conflict 1
+    //
+    // Root cause: CURL::parse discards "?" as filename
+    EXPECT_EQ("", CURL("?").GetFileName());
+
+
+    EXPECT_EQ("?", URIUtils::GetFileName("?"));
+    EXPECT_EQ("",  CURL_FileName_URIUtils_Split("?"));
```
Comment (damienruscoe) [inline xbmc/utils/test/TestURIUtils.cpp]: Can somebody explain why CURL::Parse checks for an input string "?" and skips parsing. I understand that ? could be a delimiter to CURL options. However, a local file could be named `?` and a local file cannot contain path options. IMO the CURL::Parse check and early abort should be removed and `?` should be accepted as a valid filename.

File: xbmc/utils/test/TestURIUtils.cpp [hunk 2]
```diff
@@ -600,3 +600,118 @@ TEST_F(TestURIUtils, UpdateUrlEncoding)
...
+    //
+    // Root cause: URIUtils::Split is treating a colon as a special
+    // character representing a win32 drive delimeter. This is a correct
+    // delimeter but the drive should not be returned as the "filename"
+    EXPECT_EQ("", URIUtils_Split("C:"));
+
+
+    EXPECT_EQ("C:", URIUtils::GetFileName("C:"));
+    EXPECT_EQ("", CURL_FileName_URIUtils_Split("C:"));
+    EXPECT_EQ("", URIUtils_Split("C:"));
```
Comment (damienruscoe) [inline xbmc/utils/test/TestURIUtils.cpp]: IMO `URIUtils::GetFileName("C:")` is incorrect here. This should return an empty string in this case.

File: xbmc/utils/test/TestURIUtils.cpp [hunk 3]
```diff
@@ -600,3 +600,122 @@ TEST_F(TestURIUtils, UpdateUrlEncoding)
...
+    EXPECT_EQ("/",  CURL_FileName_URIUtils_Split("\\"));
+    EXPECT_EQ("\\", URIUtils_Split("\\"));
+
+    EXPECT_EQ(".thing",  URIUtils::GetFileName("/.thing"));
+    EXPECT_EQ("/.thing", CURL_FileName_URIUtils_Split("/.thing"));
+    EXPECT_EQ("/.thing", URIUtils_Split("/.thing"));
+
+    EXPECT_EQ(".thing",  URIUtils::GetFileName("/hello/there/.thing"));
+    EXPECT_EQ(".thing", CURL_FileName_URIUtils_Split("/hello/there/.thing"));
+    EXPECT_EQ(".thing", URIUtils_Split("/hello/there/.thing"));
```
Comment (damienruscoe) [inline xbmc/utils/test/TestURIUtils.cpp]: IMO `URIUtils::GetFileName()` is returning the correct result here. The slash should prefix in the result should not be present. This appears to only appear on files in the filesystem root.

*EDIT* This comment is against line 686 where `URIUtils::Split()` is correct. The incorrect return value is on line [#681](https://github.com/xbmc/xbmc/pull/26830/files#diff-1275d8288a6a70dc38949e0357954e4b68f3f574d844abcb3edf6581161a8d58R681) and [#682](https://github.com/xbmc/xbmc/pull/26830/files#diff-1275d8288a6a70dc38949e0357954e4b68f3f574d844abcb3edf6581161a8d58R682)

File: xbmc/utils/test/TestURIUtils.cpp [hunk 4]
```diff
@@ -600,3 +600,122 @@ TEST_F(TestURIUtils, UpdateUrlEncoding)
...
+  EXPECT_EQ("udf://%2fpath%2fthing/my/archived/path",
+    URIUtils::CreateArchivePath("udf", curl, "/my/archived/path").Get());
+
+  EXPECT_EQ("iso9660://%2fpath%2fthing/my/archived/path",
+    URIUtils::CreateArchivePath("iso9660", curl, "/my/archived/path").Get());
+
+  // RAR file hostnames are __NOT__ "encoded" (%2f etc)
+  // This is a contradiction to the other file containers above
+  EXPECT_EQ("rar:///path/thing/my/archived/path",
+    URIUtils::CreateArchivePath("rar", curl, "/my/archived/path").Get());
```
Comment (damienruscoe) [inline xbmc/utils/test/TestURIUtils.cpp]: From my understanding; the path encoded in the hostname should be encoded such that it can be unambiguously parsed later. My knowledge here is limitted and I am basing my opinions on the result of ZIP and APK files. I think the RAR file should also follow the same encoding techneque.

**Comment by damienruscoe** [comment]
Just to be explicit. I do not mind fixing these issues myself. I would like to find the correct behaviour first and verify my proposed changes won't subtly break something else.

**Comment by damienruscoe** [comment]
Jenkins build this please

File: xbmc/utils/test/TestURIUtils.cpp (hunk 3 above)
Comment (damienruscoe) [reply xbmc/utils/test/TestURIUtils.cpp]: Fixed code and updated test.

File: xbmc/utils/test/TestURIUtils.cpp (hunk 2 above)
Comment (damienruscoe) [reply xbmc/utils/test/TestURIUtils.cpp]: Fixed code and updated test.

File: xbmc/utils/test/TestURIUtils.cpp (hunk 1 above)
Comment (damienruscoe) [reply xbmc/utils/test/TestURIUtils.cpp]: Fixed code according to my assumption. Updated this test. No other test cases fails.

File: xbmc/utils/test/TestURIUtils.cpp (hunk 4 above)
Comment (damienruscoe) [reply xbmc/utils/test/TestURIUtils.cpp]: Fixed code according to my assumption. Updated this test. No other test cases fails.

File: xbmc/utils/test/TestURIUtils.cpp [hunk 5]
```diff
@@ -600,3 +600,122 @@ TEST_F(TestURIUtils, UpdateUrlEncoding)
...
+  auto URIUtils_Split = [=](const std::string& in) {
+    std::string _, splitCurlFileName;
+    URIUtils::Split(in, _, splitCurlFileName);
+    return splitCurlFileName;
+  };
+  auto CURL_FileName_URIUtils_Split = [=](const std::string& in) {
+    return URIUtils_Split(CURL(in).GetFileName());
+  };
+
+  // These conflics are caused by having two implementations of the
```
[the remark this answers is not in the document]
Comment (damienruscoe) [reply xbmc/utils/test/TestURIUtils.cpp]: Comment removed.

File: xbmc/utils/test/TestURIUtils.cpp [hunk 6]
```diff
@@ -600,3 +600,122 @@ TEST_F(TestURIUtils, UpdateUrlEncoding)
...
+    EXPECT_EQ("?", URIUtils::GetFileName("?"));
+    EXPECT_EQ("",  CURL_FileName_URIUtils_Split("?"));
+    EXPECT_EQ("?", URIUtils_Split("?"));
+  }
+  {
+    // Conflict 2
+    //
+    // Root cause: URIUtils::Split is treating a colon as a special
+    // character representing a win32 drive delimeter. This is a correct
+    // delimeter but the drive should not be returned as the "filename"
```
[the remark this answers is not in the document]
Comment (damienruscoe) [reply xbmc/utils/test/TestURIUtils.cpp]: Comment removed.

**Comment by garbear** [comment]
Sorry it's only been dead computer eyes on this PR so far, but codebase fixes and especially additional test cases are something we like, so this should get some human attention at some point.

**Comment by damienruscoe** [comment]
> Sorry it's only been dead computer eyes on this PR so far, but codebase fixes and especially additional test cases are something we like, so this should get some human attention at some point.

No problem. I understand that the changes are not that exciting which usually leads to less interest :)
With not many opinions being given I have fixed the code based on what I believe should be the correct results but this will need to be introduced into an alpha release. The unit tests only proove so much.

**Comment by garbear** [comment]
Jenkins checks commit-by-commit so you can't apply patches at the end, you have to apply them commit-by-commit and force push. It's a small price to pay, but more eyeballs are drawn to PRs that pass Jenkins without noise.

**Comment by garbear** [comment]
It looks like the PR has ballooned in scope. I suggest making one small change per PR. Be detailed in the description and justify the change, because the code being touched (`Parse()`) is called from hundreds if not thousands of places, and there's a high risk in touching it.

**Comment by garbear** [comment]
This is easier to review without the overwhelming tests, but tests aren't the problem - they don't carry a maintenance burden or much risk. The PR touches a critical code path, so it better be done right - single commits for each change. For example, `CURL::Parse()` changes in their own commit. The CFile changes (if necessary) are their own commit, which comes first (as Parse() depends on it). URIUtils.cpp contains modifications to four functions - if they can't be split into separate commits, at least the single commit message has a lengthy description. Each time a critical code path is touched, test cases for that change should be in the same commit.

Above all, the PR must be bisectable, in case there's a problem we can revert individual commits. You'll notice Jenkins formats each commit - not only must we be able to build between each commit, but each commit must result in correctly formatted code (a silent jenkins) so that a single revert doesn't leave bad formatting in the codebase.

**Comment by damienruscoe** [comment]
@garbear The scope of this PR had ballooned. I have disected the PR and shared the changes across multiple other PRs. Hopefully this is down to a bitesize change.

I am well aware that `CURL` and `CURL::Parse` are part of the critical code path. I am also aware of how costly it is to call `CURL::Parse`. This is the very reason I am making these changes. Over the years of development `CURL::Parse` has been overused without the callers full realizing the cost of calling that method. This has grown into an unnecessary number of calls to Parse the same string is being executed very frequently -- Put a print statement in the `Parse()` method to see what I mean.

My aims over the next few PRs will be to reduce the quantity of superfluous calls to `Parse()` which sould give a small boost to pretty much every aspect of Kodi, including GUI navigation, TV/Movie scanning and any other place that CURL objects are used.

The reason I am starting here is because it is the critical code path :)


With your other points. 
- I will try to keep the PRs as small as possible. This should have been apparent to me already :)
- Keeping things bisectable. Any advice on how to do this? I test locally and then commit to get Jenkins to check formatting and test on other platforms. Any advice on how I can do this locally before committing?

**Comment by damienruscoe** [comment]
@garbear This branch has been minimized with the excess work spread across other PRs. I think this is ready for a review.

Thanks

**Comment by garbear** [comment]
Nice, this is much nicer to review. You have a good commit message too:

```
[CURL][tests] Added tests checking for differences between URIUtils::*

- Fixed issue found in testing when a service is unavailable. Check the service is non-null before calling its methods.
- Fixed URIUtils::Split for absolute filename at the fs root.
- Fixed CURL parsing. A question mark may be a valid filename.
- Fixed inconsistency when getting filename from a windows drive spec.
- Fixed encoding of rar files to be consistent with zip etc...
```

Can these be split into individual commits? It's not clear which fix goes with which change.

If a commit fixes something, make sure the word fix appears in the commit title.

**Comment by damienruscoe** [comment]
> Nice, this is much nicer to review. You have a good commit message too:

Thanks. I agree. From the standpoint of a reviewer it was too much. I was overzelous with the amount of wasted cycles I saw.

> Can these be split into individual commits? It's not clear which fix goes with which change.

OK. Ill split this further.

**Comment by damienruscoe** [comment]
Jenkins build this please

**Review by garbear — APPROVED** [review]

File: xbmc/utils/URIUtils.cpp [hunk 7]
```diff
@@ -295,6 +295,7 @@ bool URIUtils::HasParentInHostname(const CURL& url)
 {
   return url.IsProtocol("zip") || url.IsProtocol("apk") || url.IsProtocol("bluray") ||
          url.IsProtocol("udf") || url.IsProtocol("iso9660") || url.IsProtocol("xbt") ||
+         url.IsProtocol("rar") ||
```
[the remark this answers is not in the document]
Comment (garbear) [reply xbmc/utils/URIUtils.cpp:298]: out of scope

**Comment by garbear** [comment]
Copilot is right about insufficient tests, and tests are so cheap to generate, you might as well cover edge cases in the code. But keep in mind a human reads all the code, so don't go too crazy.
